circuit/environment/src/helpers/r1cs.rs

```diff
@@ -91,9 +91,22 @@ impl<F: PrimeField> R1CS<F> {
         self.counter.add_constraint(constraint);
     }
 
-    /// Returns `true` if all constraints in the environment are satisfied.
+    /// Returns `true` if both
+    /// - all constraints in the environment are satisfied.
+    /// - all constraints use variables corresponding to the declared variables
     pub fn is_satisfied(&self) -> bool {
-        self.constraints.iter().all(|constraint| constraint.is_satisfied())
+        let constraints_satisfied = self.constraints.iter().all(|constraint| constraint.is_satisfied());
+        let vars_consistent = self.constraints.iter().all(|constraint| {
+            let (a, b, c) = constraint.to_terms();
+            [a, b, c].into_iter().all(|m| {
+                m.to_terms().into_iter().all(|(var, _)| match var {
+                    Variable::Constant(_value) => false, // terms should only contain private/public
+                    Variable::Private(index, value) => self.private[*index as usize].value() == **value,
+                    Variable::Public(index, value) => self.public[*index as usize].value() == **value,
+                })
+            })
+        });
+        constraints_satisfied && vars_consistent
```

Comment:
if we go with the early return, there is no need for the `vars_consistent` binding or this final line